frontend/packages/dev-console/src/components/import/pipeline/__tests__/pipeline-template-utils.spec.ts

```diff
@@ -0,0 +1,291 @@
+      pipeline: {
+        enabled: true,
+        template: pipelineTemplate,
+      },
+    };
+    return formData;
+  };
+
+  it('should create an almost empty pipeline for a template with only task data (empty task)', async () => {
+    const pipelineTemplate: Pipeline = {
+      spec: {
+        tasks: [],
+      },
+    };
+
+    await createPipelineForImportFlow(createFormData(pipelineTemplate));
+
+    const expectedPipeline = {
+      metadata: {
+        name: 'an-app',
+        namespace: 'a-project',
+        labels: { 'app.kubernetes.io/instance': 'an-app' },
+      },
+      spec: {
+        params: undefined,
+        tasks: [],
+      },
+    };
+
+    expect(k8sCreate).toHaveBeenCalledTimes(1);
+    expect(k8sCreate).toHaveBeenCalledWith(PipelineModel, expectedPipeline, {
+      ns: 'a-project',
+    });
+    expect(createPipelineResource).toHaveBeenCalledTimes(0);
+  });
+
+  it('should create a pipeline for a template with params, resources, workspaces, tasks (all empty)', async () => {
+    const pipelineTemplate: Pipeline = {
+      spec: {
+        params: [],
+        resources: [],
+        workspaces: [],
+        tasks: [],
+        serviceAccountName: 'service-account',
+      },
+    };
+
+    await createPipelineForImportFlow(createFormData(pipelineTemplate));
+
+    const expectedPipeline = {
+      metadata: {
+        name: 'an-app',
+        namespace: 'a-project',
+        labels: { 'app.kubernetes.io/instance': 'an-app' },
+      },
+      spec: {
+        params: [],
+        resources: [],
+        workspaces: [],
+        tasks: [],
+        serviceAccountName: 'service-account',
+      },
+    };
+
+    expect(k8sCreate).toHaveBeenCalledTimes(1);
+    expect(k8sCreate).toHaveBeenCalledWith(PipelineModel, expectedPipeline, {
+      ns: 'a-project',
+    });
+    expect(createPipelineResource).toHaveBeenCalledTimes(0);
+  });
+
+  it('should create a pipeline for a template with filled params, resources, workspaces and tasks', async () => {
+    const pipelineTemplate: Pipeline = {
+      spec: {
+        params: [{ name: 'a-param', default: 'default value', description: 'a description' }],
+        resources: [{ type: 'resource-type', name: 'a-resource' }],
+        workspaces: [{ type: 'workspace-type', name: 'a-workspace' }],
+        tasks: [
+          {
+            name: 'build',
+            taskRef: {
+              name: 'a-task',
+            },
+          },
+        ],
+      },
+    };
+
+    await createPipelineForImportFlow(createFormData(pipelineTemplate));
+
+    const expectedPipeline = {
+      metadata: {
+        name: 'an-app',
+        namespace: 'a-project',
+        labels: { 'app.kubernetes.io/instance': 'an-app' },
+      },
+      spec: {
+        params: [{ name: 'a-param', default: 'default value', description: 'a description' }],
+        resources: [{ type: 'resource-type', name: 'a-resource' }],
+        workspaces: [{ type: 'workspace-type', name: 'a-workspace' }],
```

Comment:
Oooh, also can I get you to type all the test data in the file? I think it'll help when we apply the above change.

Reply:
@andrewballantyne Good catch. I added a `: Pipeline` type check also to the exectedPipeline and removed the invalid type from the unit tests. This was required by the PipelineWorkspace before. After changing the PipelineWorkspace `type PipelineWorkspace = Param;` some other type definitions fails.

This happen because the PipelineWorkspace was also used to save the formik workspace state of the Start Pipeline modal. I added a new type for the formik data and fixed the PipelineWorkspace and PipelineTaskWorkspace definitions. So this commit is now a little bigger then initial. Wdyt?

Reply:
The size increase mostly was type handling - if not entirely type handling... that kinda size bloat is not really possible to be risky. It should fail on your PR if it was invalid, and it in itself cannot cause a bug since it doesn't exist in the bundled js that is run in the cluster.

I'm okay with this, I'd like to try out the UI once before giving an approval, but the code looks good 👍  Great tests again.